scene/resources/material.cpp

```diff
@@ -985,7 +1008,13 @@ void BaseMaterial3D::_update_shader() {
 	if (flags[FLAG_ALBEDO_FROM_VERTEX_COLOR]) {
 		code += "\talbedo_tex *= COLOR;\n";
 	}
-	code += "\tALBEDO = albedo.rgb * albedo_tex.rgb;\n";
+
+	// Use particle flipbook animation if specified.
+	code += "\tif (particles_anim_blend) {\n";
+	code += "\t\tALBEDO = albedo.rgb * mix(albedo_tex.rgb, texture(texture_albedo, particles_anim_next_frame_uv_coord).rgb, particles_anim_next_frame_mix);\n";
+	code += "\t} else {\n";
```

Comment:
You should read and store the ``albedo_blend_tex``  here. This will allow you to only read the texture once. Below when reading alpha, you can just use ``albedo_blend_tex.a``.

```suggestion
	code += "\tvec4 albedo_blend_tex = vec4(1.0);\n";
	code += "\tif (particles_anim_blend) {\n";
	code += "\t\talbedo_blend_tex = texture(texture_albedo, particles_anim_next_frame_uv_coord);\n";
	code += "\t\tALBEDO = albedo.rgb * mix(albedo_tex.rgb, albedo_blend_tex.rgb, particles_anim_next_frame_mix);\n";
	code += "\t} else {\n";
```